kaniko-build docs (#6040)
kyma-project/test-infra: development/kaniko-build/README.md

# kaniko-build

> This tool serves as an intelligent wrapper for `kaniko-project/executor`. It's built to reduce the complexity of building the Docker images and remove the need of using Docker in Docker when building images in K8s infrastructure.

Reviewer: ```suggestion
This tool serves as an intelligent wrapper for `kaniko-project/executor`. It reduces the complexity of building Docker images and removes the need of using Docker in Docker when building images in K8s infrastructure.
```

Key features:
* Automatically provide a default tag, which is computed based on a template provided in `config.yaml`
* ~~Allow for concurrent builds of image variants that use the same `Dockerfile`~~ See Known issues #1
* Support for adding multiple tags to the image
* Save the `kaniko-project/executor` command outputs to separate files
* Support for pushing images to different repositories with different tags when running in Prow's presubmit job
* Support for pushing the same images to multiple repositories
* Supports caching of built layers to reduce build times

## Known issues

> 

Author: It matters unfortunately, because I'm referencing to it in the 7th line of the docs until the bug is resolved.

> 

Reviewer 2: Good point! :D

## Use config.yaml file

`kaniko-build` requires a configuration file to be provided with a set of variables, which are used during the execution.
A `--config` flag is required.

For more information, refer to the [config.go](./config.go) file.

Example file:
```yaml
registry: eu.gcr.io/kyma-project
reproducible: true
log-format: json
cache:
  enabled: true
  cache-repo: eu.gcr.io/sap-kyma-neighbors-dev/cache
  cache-run-layers: true
```

## Build multiple variants of the same image

The tool also includes the functionality of re-using the same `Dockerfile` to concurrently build different variants of the same image.
`variants.yaml` is the file that contains pre-defined set of the same `ARG` substitutions with different values.
It's very handy when there is a requirement to build an image with different versions of the same binary, for example, for different versions of Kubernetes or Go.

The file has a simple structure:
```yaml
'main':
  KUBECTL_VERSION: "1.24.4"
'1.23':
  KUBECTL_VERSION: "1.23.9"
```

> To use this feature, make sure that:

Reviewer 2: This looks like a prerequisite for using the `variants.yaml` file. If yes, it needs to go in front of the instructions how to use that, because first the prerequisites need to be fulfilled before users can start using it.

So, for example, after line 41 (just guessing), I'd slip in:

> To use this feature, fulfill the following prerequisites:
> 
> * The `variants.yaml` file is stored in the **same directory** as the `Dockerfile`.
> * Your `Dockerfile` contains `ARG` directives that are named after keys in `variants.yaml`.

Author: I think that this place is also fine. Because first line is the description of the functionality. I've formatted it as simple paragraph, rather than prerequisites in mind.

> * You have the `variants.yaml` file in the **same directory** as the `Dockerfile` file.
> * Your `Dockerfile` contains `ARG` directives which are named after keys in `variants.yaml`

Reviewer: ```suggestion
* you have the `variants.yaml` file in the **same directory** as the `Dockerfile`
* your `Dockerfile` contains `ARG` directives which are named after keys in `variants.yaml`
```

Reviewer: These changes will result in greater consistency of the text.

> ## Usage

Reviewer 2: What does "Usage" mean here? Is this simply a code example?  Can we find a more meaningful headline?

Author: I don't think so. It's just usage output of `kaniko-build -h`. I don't see sense in adding more complex heading here.
